### ISSUE TITLE
Addin to load target into global environment?

### ISSUE THREAD

**Issue by MilesMcBain** [ropensci/drake#763]
When developing my `drake` pipeline, I found myself frequently typing:

`target <- readd(target)`

to inspect a recently built target and then use it interactively to build out the next piece of the pipeline that uses that target. I wrote myself a little [elisp function](https://github.com/MilesMcBain/spacemacs_cfg/blob/windows/private/ess/packages.el#L362) to do this for me for the object currently under the cursor. It feels pretty neat. I thought it might make a nice RStudio addin to ship with `drake`.

Here's an implementation that I'd be happy to PR if you like the idea:

```
 rs_get_symbol_at_cursor <- function(){
  context <- rstudioapi::getActiveDocumentContext()
  if(context$id == "#console") return(NULL)
  
  cursor_pos <- context$selection[[1]]$range$start
  cursor_line <- cursor_pos[1]
  cursor_column <- cursor_pos[2]
  
  r_symbol_pattern <- "[.A-Za-z][.A-Za-z0-9_]+"
  
  line_symbols <- gregexpr(text = context$contents[cursor_line],
                           pattern = r_symbol_pattern)
  
  match_starts <- line_symbols[[1]]
  match_ends <- attr(x = line_symbols[[1]], "match.length")
  match_index <- which(cursor_column >= match_starts & cursor_column <= match_ends)
  
  the_symbol <- substr(context$contents[cursor_line],
                       start = match_starts[match_index],
                       stop = match_ends[match_index])
  the_symbol
}

readd_at_cursor_addin <- function(){
  target <- rs_get_symbol_at_cursor()
  if(!is.null(target)){
    message(glue::glue("Loading {target} from cache to global evironment."))
    eval(parse(text = glue::glue("{target} <- drake::readd({target})")),
         envir = .GlobalEnv)
  }
}
```

**Comment by wlandau** [comment]
Cool, thanks Miles! I will try it out.

By the way, you can use `drake_build()` and `drake_debug()` to build individual targets. Both call `loadd(deps = TRUE)`, which loads all the dependencies of the supplied targets.

If the session running `make()` is different from the session you use to inspect the results, you might also like `loadd(lazy = bind)`, which uses active bindings.

**Comment by wlandau** [comment]
@MilesMcBain, a PR for the addin would be amazing.

When I tried it out, I ended up reworking the code somewhat. Please see my comments below.

```r
rs_addin_loadd <- function(){
  target <- rs_get_symbol_at_cursor()
  if (is.null(target)) {
    return()
  }
  cache <- get_cache()
  message(
    "Loading target ",
    shQuote(target),
    " into global evironment.\nCache: ",
    # By default, drake searches parent directories for the project's cache,
    # so it may good to tell the user which cache is being used.
    shQuote(cache$driver$path)
  )
  # I think we can call loadd() directly without eval(parse()).
  loadd(list = target, envir = globalenv(), cache = cache)
}

rs_get_symbol_at_cursor <- function() {
  # I would prefer rstudioapi to be an optional package.
  assert_pkg("rstudioapi")
  context <- rstudioapi::getActiveDocumentContext()
  if(identical(context$id, "#console")) {
    return(NULL)
  }
  cursor_pos <- context$selection[[1]]$range$start
  cursor_line <- cursor_pos[1]
  cursor_column <- cursor_pos[2]
  r_symbol_pattern <- "[.A-Za-z][.A-Za-z0-9_]+"
  line_symbols <- gregexpr(
    text = context$contents[cursor_line],
    pattern = r_symbol_pattern
  )
  match_starts <- line_symbols[[1]]
  # I needed to add match_starts to get the correct match_ends.
  match_ends <- match_starts + attr(x = line_symbols[[1]], "match.length") - 1
  match_index <- which(
    cursor_column >= match_starts &
    cursor_column <= match_ends
  )
  # When it comes to long argument lists, I make a new indented line for each argument.
  # This differs from the norm, but I do have reasons.
  # 1. It is easier to keep lines under 80 characters.
  # 2. My eyes can more easily track the text if the indentation
  #    does not depend on the length of the function name.
  #    I remember this principle coming up in an unrelated conversation
  #    with a graphic designer a while back.
  substr(
    context$contents[cursor_line],
    start = match_starts[match_index],
    stop = match_ends[match_index]
  )
}
```

Other remarks:
- I would prefer this code to live in a new `R/addin-loadd.R` file with something like this for `addins.dcf`:

```
Name: loadd
Description: Load a drake target
Binding: rs_addin_load
Interactive: true
```

- I am brand new to RStudio addins. What degree of automated testing is available? Maybe we could use a `debug` or `test` argument to drop the interactivity requirement? If we do have the unreachable lines of code after that, I think we could wrap them in `# nocov` and add reasonable justification in other comments. I am not sure what the community standard is here.
- Could addins also cover argument autocompletion? Candidates: `loadd()`, `readd()`, `clean()`, `diagnose()`, `drake_build()`, `r_drake_build()`, and `drake_debug()`.

**Comment by MilesMcBain** [comment]
Okay great! 

I quite like your point Re argument lists, I may start using this myself. The other changes make sense to me.

There's no testing hooks that I know of for addins. In the past I've tried to pull out the important logic into separate testable functions that the addin function wraps as thinly as possible. So we could go that way.

Or there's this way I just thought of:
```r
rs_get_symbol_at_cursor(context =  rstudioapi::getActiveDocumentContext()){
...
}
```

Then we can have some tests that mock up context objects. We'd need at least one more wrapper though if you want to `assert_pkg("rstudioapi")`

**Comment by MilesMcBain** [comment]
Okay wait my way 'I just thought of' is just a slightly more streamlined version of what I have always done. The novel thing for me was mocking the context object. I hope that's not confusing.

**Comment by wlandau** [comment]
Great idea about mocking context objects. As for `rstudoiapi`, maybe we could do something like this:

```r
rs_get_symbol_at_cursor(context = NULL){
  assert_pkg("rstudioapi")
  context <- context %||% rstudioapi::getActiveDocumentContext()
  # ...
}
```

[`%||%` and `%|||%` are already defined in `R/utils-utils.R`](https://github.com/ropensci/drake/blob/968ff1e42ee84dfdc1f0b92c7feac16683eb5ea1/R/utils-utils.R#L2-L16). I think that gets us appropriately `covr`'ed.

**Comment by wlandau** [comment]
Implemented by @MilesMcBain in #825.
